### ISSUE TITLE
Cannot create entities after deserializing the world

### ISSUE THREAD

**Issue by dacete** [genaray/Arch.Extended#29]
I am trying to use Persistence in order to send the world state over to another client and continue simulation there. However, I am getting this error when trying to create entities after deserialization.

```
ArrayTypeMismatchException: Attempted to access an element as a type incompatible with the array.
System.Array.Copy (System.Array sourceArray, System.Int32 sourceIndex, System.Array destinationArray, System.Int32 destinationIndex, System.Int32 length) (at <59a524f39437477884870e130f2eaa4f>:0)
System.Array.Resize[T] (T[]& array, System.Int32 newSize) (at <59a524f39437477884870e130f2eaa4f>:0)
Arch.Core.JaggedArray`1[T].EnsureCapacity (System.Int32 capacity) (at <f04907b0740a41dcb23e49fd4f7f024b>:0)
Arch.Core.EntityInfoStorage.EnsureCapacity (System.Int32 capacity) (at <f04907b0740a41dcb23e49fd4f7f024b>:0)
Arch.Core.World.GetOrCreate (System.Span`1[T] types) (at <f04907b0740a41dcb23e49fd4f7f024b>:0)
Arch.Core.World.Create[T0] (T0& t0Component) (at <f04907b0740a41dcb23e49fd4f7f024b>:0)`
```
This is enough to reproduce. Just create the world, serialize it and deserialize it, and when trying to create, it errors.

```
public struct Test
{
    public int value;
}        

ArchBinarySerializer serializer = new ArchBinarySerializer();
World = World.Create();
var bytes = serializer.Serialize(World);
World = serializer.Deserialize(bytes);
World.Create(new Test { value = 2 });
```
I also used ComponentRegistry but to no avail.
One interesting behaviour, if I add the entity before serializing, the error is fired from the .Deserialize line instead. Here is that stacktrace, if it helps.

```
ArrayTypeMismatchException: Attempted to access an element as a type incompatible with the array.
System.Array.Copy (System.Array sourceArray, System.Int32 sourceIndex, System.Array destinationArray, System.Int32 destinationIndex, System.Int32 length) (at <59a524f39437477884870e130f2eaa4f>:0)
System.Array.Resize[T] (T[]& array, System.Int32 newSize) (at <59a524f39437477884870e130f2eaa4f>:0)
Arch.Core.JaggedArray`1[T].EnsureCapacity (System.Int32 capacity) (at <f04907b0740a41dcb23e49fd4f7f024b>:0)
Arch.Core.EntityInfoStorage.EnsureCapacity (System.Int32 capacity) (at <f04907b0740a41dcb23e49fd4f7f024b>:0)
Arch.Core.Extensions.Dangerous.DangerousWorldExtensions.EnsureCapacity (Arch.Core.World world, System.Int32 capacity) (at <f04907b0740a41dcb23e49fd4f7f024b>:0)
Arch.Persistence.WorldFormatter.Deserialize (MessagePack.MessagePackReader& reader, MessagePack.MessagePackSerializerOptions options) (at <af3ea2856d7b42a2bd70a4676dc1a254>:0)
MessagePack.MessagePackSerializer.Deserialize[T] (MessagePack.MessagePackReader& reader, MessagePack.MessagePackSerializerOptions options) (at <e0125fc189a34138b45ca77ebe60d374>:0)
Rethrow as MessagePackSerializationException: Failed to deserialize Arch.Core.World value.
MessagePack.MessagePackSerializer.Deserialize[T] (MessagePack.MessagePackReader& reader, MessagePack.MessagePackSerializerOptions options) (at <e0125fc189a34138b45ca77ebe60d374>:0)
MessagePack.MessagePackSerializer.Deserialize[T] (System.ReadOnlyMemory`1[T] buffer, MessagePack.MessagePackSerializerOptions options, System.Threading.CancellationToken cancellationToken) (at <e0125fc189a34138b45ca77ebe60d374>:0)
Arch.Persistence.ArchBinarySerializer.Deserialize (System.Byte[] world) (at <af3ea2856d7b42a2bd70a4676dc1a254>:0)
```

I got the Persistence package from https://www.nuget.org/packages/Arch.Persistence, is it maybe outdated? I see there's tag 1.0.8 on github, but I'm using Unity, and this is how I got it to compile.

**Comment by genaray** [comment]
Which version did you use? ^^
1.0.2 is the most recent one.

Im gonna check that real quick, looks like its a bug.

**Comment by genaray** [comment]
I did test that with the latest persistence version rn and it worked fine...
What version are you using? :0

**Comment by dacete** [comment]
I'm on 1.0.2 too. and 1.2.6.8-alpha for Arch. 
Later today I'll try to repro with a fresh project.

**Comment by genaray** [comment]
> I'm on 1.0.2 too. and 1.2.6.8-alpha for Arch. Later today I'll try to repro with a fresh project.

Alright would be great if you could lemme know if it works or not ^^

**Comment by dacete** [comment]
So I created a new unity project, using version 2023.2.0b12. Then I added Arch and Arch.Persistence, here are the version numbers
![image](https://github.com/genaray/Arch.Extended/assets/9105354/e8492857-0976-4f42-9f12-ff6d7b65efcc)
I wrote the same script as above, and it still happens. 
Could it be because I ran it in unity? Let me know if I can provide anything else. I'm also in the discord, under username noice.

**Comment by genaray** [comment]
> So I created a new unity project, using version 2023.2.0b12. Then I added Arch and Arch.Persistence, here are the version numbers ![image](https://user-images.githubusercontent.com/9105354/274049942-e8492857-0976-4f42-9f12-ff6d7b65efcc.png) I wrote the same script as above, and it still happens. Could it be because I ran it in unity? Let me know if I can provide anything else. I'm also in the discord, under username noice.

Thats incredibly weird... i can not reproduce this issue. However I'm not using unity, just plain c# and also i tried it with the repo source-code and not the nugget. Im gonna setup a new project later today and see if the nugets is the problem ^^

**Comment by dacete** [comment]
I made a C# console project, and it worked there. Is it unity-related then? Not sure what I could check next

**Comment by genaray** [comment]
> I made a C# console project, and it worked there. Is it unity-related then? Not sure what I could check next

In this case, it MUST be Unity-related. But why? This makes no sense at all. Do you have code stripping on extreme? Can it be that the reflection-code is stripped somewhere? :o

**Comment by dacete** [comment]
Managed to repro out of unity, if I made a new C# project and set the framework to .NET Core 3.1 it throws this error. Unity uses .NET Standard 2.1 and I read somewhere it's comparable to Core 3.0

I know NuGet can provide netstandard2.1, net 6 and net 7 packages. Maybe it works only in the latter 2?

**Comment by genaray** [comment]
Arch.Extended supports all three. But not Net.Core 3.1 (at least not directly). That would also explain the error.

As I understand it, NetCore 3.1 is dying anyway and will soon no longer be supported. Unity doesn't use .NetCore 3.1 either though... but .Net Standard 2.1 ^^

**Comment by dacete** [comment]
That's fair enough, but I'm also getting the error in Net Standard 2.1, in Unity. I couldn't find an option for Net Standard 2.1 when I made the C# project, were you able to test on a project with Net Standard 2.1?

**Comment by dacete** [comment]
So I went back to my (Net 7.0?) console project, and instead of using NuGet, I copied the DLLs that I had in Unity, basically the NuGet netstandard2.1 DLLs. The error happens then as well. Seems to me that netstandard2.1 DLLs coming from NuGet have this bug. Can you repro?

**Comment by genaray** [comment]
> That's fair enough, but I'm also getting the error in Net Standard 2.1, in Unity. I couldn't find an option for Net Standard 2.1 when I made the C# project, were you able to test on a project with Net Standard 2.1?

NetStandard is more like a "promise". You can not make a project purely NetStandard 2.1, you can just promise that it uses the API. Unfortunately, i do not have time to test it with Unity rn. 

Just a small sidenote. Have you tested to include the source and debug it to see where exactly the serializer fucks up? ^^

**Comment by dacete** [comment]
[Arch.zip](https://github.com/genaray/Arch.Extended/files/12895343/Arch.zip)
So I'm using Visual Studio 2022 for this, I make a new Console project, and right click on dependencies > Add project reference

**Comment by dacete** [comment]
About the source code, I tried I think at the start, but got errors since it was made with a newer version of C#, maybe because file-scope namespaces? Don't remember exactly, but I gave up and just used NuGet.

**Comment by genaray** [comment]
> [Arch.zip](https://github.com/genaray/Arch.Extended/files/12895343/Arch.zip) So I'm using Visual Studio 2022 for this, I make a new Console project, and right click on dependencies > Add project reference

Does it work for you with those dlls? 
I created a new project, included the provided dlls and it still annoys the crab out of me that more dlls are missing. E.g. Messagepack.Annotations xD

**Comment by dacete** [comment]
Hmm, I might've been too hasty, I imported all the nuget packages that I had in unity and I can't repro the bug in console app. 
In the end I ended up switching to Flax Engine, and the same code works there no problem (using net 7 dlls), so this is not as important to me anymore. Best way to repro would probably be to create a unity project, but I don't blame you if you don't fix this, since it might be related to unity and not your code.
